test/diag2dwd.jl

```diff
@@ -217,6 +218,317 @@ end
+    rows = columns .|> (x -> acset[x]) |> (x -> zip(x...)) |> collect
+    rem_parts!(acset, table,
+      setdiff(parts(acset, table),
+        unique(i -> rows[i], eachindex(rows))))
+    return acset
+  end 
+
+  """
+      function unique_by(acset, column_names::Vector{Symbol})
+
+  Given column names from the same table, return a copy of the acset with
+  duplicate rows removed. Removal of rows is performed with prejudice.
+
+  WARNING: This function does not check if other tables index into the one
+  given. Removal of rows is performed with prejudice.
+
+  See also: [`unique_by!`](@ref).
+
+  # Examples
+  ```julia-repl
+  julia> g = @acset Graph begin
+    V = 2
+    E = 3
+    src = [1,1,2]
+    tgt = [1,1,2]
+  end
+  julia> unique_by(g, :E, [:src, :tgt])
+  ``` 
+  """
+  function unique_by(acset, table::Symbol, columns::Vector{Symbol})
+    acset_copy = copy(acset)
+    unique_by!(acset_copy, table, columns)
+  end 
+
+  """
+      function compose_decapodes(decapodes_vars::Vector{Tuple{NamedDecapode{
+        Any, Any, Symbol}, Vector{Symbol}}}, relation::RelationDiagram)
+
+  Compose a list of decapodes as specified by the given relation diagram.
+
+  The decapodes must be given in the same order as they were specified in the
+  relation.
+
+  State variables (such as the (C,V) given in the head of the following
+  @relation) do not affect the result of a composition.
+
+  # Examples
+  ```julia-repl
+  julia> compose_diff_adv = @relation (C,V) begin
+    diffusion(C, ϕ₁)
+    advection(C, ϕ₂, V)
+    superposition(ϕ₁, ϕ₂, ϕ, C)
+  end
+  julia> compose_decapodes([(Diffusion, [:C, :ϕ]), (Advection, [:C, :ϕ, :V]),
+    (Superposition, [:ϕ₁, :ϕ₂, :ϕ, :C])], compose_diff_adv)
+  ````
+  """
+  function compose_decapodes(decapodes_vars::Vector{Tuple{NamedDecapode{
+    Any, Any, Symbol}, Vector{Symbol}}}, relation::RelationDiagram)
+    # TODO: Calling this function "compose" is something of a misnomer.  e.g.
+    # You could very well give a vector containing a single decapode. Then the
+    # effect of applying the relation would be something like renaming
+    # variables. So this function should really be called oapply_decapodes (or
+    # just oapply if you are willing to import oapply from
+    # Catlab.WiringDiagrams).
+    r = relation
+    copies = decapodes_vars .|> first .|> copy
+    # TODO: We should also check that the number of variables given in the
+    # relation is the same as the number of symbols in the corresponding vector
+    # of Vars.
+
+    # Step -4: Determine the mapping of global ports to local ports.
+    # In a RelationDiagram, this is baked into the order of rows in the Port
+    # table.
+    # (i.e. for each (box,junction) pair, determine whether this corresponds to
+    # the 1st Var of that box, or the 2nd Var of that box, etc.)
+    # tuples (box, indices of rows with that box):
+    box_rows = [(b, incident(r, b, :box)) for b in unique(r[:box])]
```

Comment:
The schema for UWDs is 

```julia
@present SchUWD(FreeSchema) begin
  Box::Ob
  Port::Ob
  OuterPort::Ob
  Junction::Ob

  box::Hom(Port, Box)
  junction::Hom(Port, Junction)
  outer_junction::Hom(OuterPort, Junction)
end
```

taking `unique(r[:box])` is asking for the unique boxes based on the list of box for each port. This is allowed to have duplicate entries because if a box has multiple ports it will show up once for each port on that box. But if you just map over `parts(r, :Box)`, then you will get each box exactly once.

The function `incident(r, b, :box)` where b is a box ID,  gives you the ports that live on that box in the order they are defined in the relation macro. You shouldn't have to do any other indexing arithmetic on those values. You can do 

```julia
for (i, p) in enumerate(incident(r, b, :box))
```

to get the local port number (i) and the global port number (p) for every port on the box (b). If you stick that in another loop over b, then you get it for all the ports in the system.

```julia
for b in parts(r, :Box)
  for (i, p) in enumerate(incident(r, b, :box))
     # b is the box
     # i is the local port id
     # p is the global port id
  end
end
```